### ISSUE TITLE
[DOCS] Provide high-level guidance for testing HTTP REST client

### ISSUE THREAD

**Issue by jrodewig** [elastic/elasticsearch#55258]
Outside of the [Java testing framework docs][0], which is intended for plugin/integration authors, we don't currently provide much guidance on how application developers can test the HTTP REST client.

### Proposal
Because we can't realistically provide guidance for all the languages/eco-systems/CI environments that exist, we should keep guidance high-level.

For example, we can point uses to our [docker containers][1], which can be used to spin up clusters in docker-supported environments. However, we probably want to avoid build-specific guidance.

[0]: https://www.elastic.co/guide/en/elasticsearch/reference/master/testing-framework.html
[1]: https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#docker-cli-run-dev-mode

Relates to #55257

**Comment by PhaedrusTheGreek** [comment]
It would be helpful if we could supply samples of how to spawn ES from Gradle and Maven at least.

Here are some I found:  https://github.com/joel-costigliola/elastic-search-test

**Comment by rfan-debug** [comment]
Is it possible to get an ES-server emulator for testing purpose?

**Comment by adamtaylor13** [comment]
Hey all, I'm curious as I've been dealing with this frequently over the past few months. Is there a better way to test that certain prebuilt queries will match an expected set of documents? We typically dynamically build an ES query off of internal filters our applications creates, and it would be a huge improvement if we could simply verify that a given filter (our model) will be built into an ES query which matches some set of documents.

I guess ideally I'm thinking of some way to test queries built by QueryBuilder against some rules engine? Whatever ES uses internally to score/match documents. This would prevent us from needing to spin up an entire ES instance, manually insert documents, and then actually fire a REST request against this test instance, just to verify that our queries are matching as expected.

Sorry if this is too specific. It seemed like it fell under the general advice of "ES Testing Patterns".

**Comment by dadoonet** [comment]
Just a note about this as it may be useful for readers. Here's what I'm often saying [on discuss](https://discuss.elastic.co/t/in-memory-testing-with-resthighlevelclient/106196/6).

To do integration tests, you can:

* <strike>Use the Gradle tools elasticsearch already has. You can read some information about this here: https://github.com/elastic/elasticsearch/issues/21119</strike>
* Use the Maven plugin: https://github.com/alexcojocaru/elasticsearch-maven-plugin
* Use Ant scripts like http://david.pilato.fr/blog/2016/10/18/elasticsearch-real-integration-tests-updated-for-ga/
* Using Docker and TestContainers: https://www.testcontainers.org/modules/elasticsearch/
* Using Docker from maven: https://github.com/dadoonet/fscrawler/blob/e15dddf72b1ed094dad279d492e4e0314f73683f/pom.xml#L241-L289 
* Also this sample project shows how to use Elasticsearch Test Classes: https://github.com/joel-costigliola/elastic-search-test

**Comment by mark-vieira** [comment]
I'd probably suggest some Docker-based solution here. Using TestContainers as listed by @dadoonet is a great way to do this.

**Comment by odinho** [comment]
The common foot guns would be good to keep in.

As in using any time with 'now' in the query is likely to make for flaky tests as 'now' is impure.
You should *really* clear the _entire_ index (not just the document IDs you know about, since elastic might store more things?) between runs, since just one document being left will influence the scoring. Etc.

I'm having trouble making the ranking deterministic as you can see ^_^

**Comment by dakrone** [comment]
Closing this as we've removed the high level rest client in favor of the [Java client](https://github.com/elastic/elasticsearch-java/).

**Comment by odinho** [comment]
I'd like to reopen this as the Rest API very much still exist. That's what we are using from Node.js. So basically this is about every other language than Java, which it sounds like have this ordered already.

I'm basically just requesting some best practices, a high level document describing how you can write tests utilising elastic search for testing ranking and similar.

We still struggle with this. The tests are flaky (we haven't yet made filling in elastic before the integration tests fully deterministic).

**Comment by mltsy** [comment]
I agree that this should be re-opened.  I'm also using the Node.js client, and looking for best practices related to integration testing with Elasticsearch.

For instance: Is there a mock library? Can I run Elastic in memory, or should I just spin up a docker container?  Is there a special image that makes testing use-cases easier?  Is there an easy/recommended way to load a snapshot or otherwise seed the cluster for testing? Are all of these just things nobody has really figured out, so I need to figure them out and share my own experience?

**Comment by mltsy** [comment]
A colleague just pointed me to this [ES mock library](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-testing.html#_elasticelasticsearch_mock), which may be helpful for those using the JS client... 🤔

**Comment by pioorg** [comment]
Hello!
I'm pleased to see this discussion happening ;-)
Context: I've been working for Elastic for half a year now, as dev adv. Previously, for years, my job was backend developer, trainer, sometimes consultant. 80% Java stack. Java & Testcontainers Champion.
Frankly, I was surprised, that in general I couldn't easily find a guideline / policy of testing using _real_ Elasticsearch. That's why for some of my talks (and code demos) I've developed my own. (Which proves exactly nothing, but that I wasn't able to find it ;-) )

TL;DR: things mentioned here: initialising with data, snapshots and restores, are possible with ES in (automated) tests. However, they could be easier to find and use.

First, in grand scheme of things, Elasticsearch can be used in as a real external dependency (just like databases, message queues, etc.) for development purposes, automated dev workstations tests and CI/CD pipelines, using the set of orchestrating libraries called Testcontainers, which @dadoonet already mentioned.
Basically, such a library calls Docker under the hood, takes care or proper spin up and (perhaps more important) tear down and releasing resources, can copy files back and forth, call command in the container, show the logs, etc. In other words: instead of starting & stopping manually in terminal/CLI/script or Docker Desktop, tests can start _real dependencies_ in their `@BeforeAll` setup phase.

Secondly, after we have the real dependency up and running (in our case Elasticsearch), there's the issue of priming it with data and (re)storing its state to the desired one. And here we come to the point which (the way I see it) is the main concern of this thread.
Naive approach is to do that manually, and I see it as the worst, because it's really inefficient. Although it's a decent starting point.
Better approach is to have everything [(mappings & actual data in njson)](https://github.com/pioorg/Testcontainers-performance-up/tree/c10cb6ea74e0745058b9e862739cf7b54391c540/src/main/resources/config/elasticsearch) ready at the host machine, and make automated calls, and do bulk inserts or updates (ideally copy the files to the container _exactly once_ (because network penalty)), then drop indicies and prime with data again before each test. In my example I'm using `curl` instead of Java client (which could make more sense at the first glance), solely because `curl` can be used regardless of the stack, so e.g. Go or Node.js developers can use it too ;-)
If someone really needs to hunt the performance (and for any sophisticated code base this happens sooner or later), the idea is to [snapshot](https://github.com/pioorg/Testcontainers-performance-up/blob/f33c75edfe9764ea2dfa9fd62d85b0841821fca2/src/test/java/dev/softwaregarden/tcPerfUp/misc/ElasticsearchContainerHelper.java#L49) Elasticsearch after priming with data and restore the snapshot before each test. I find this idea really efficient tech-wise, but for some engineers it's not instantly obvious. Maybe because it takes a bit more for ES to run snapshots ;-)
Such a snapshot can be even copied back to the host machine, for evaluation or caching.

Another approach (popular for some SQL DBs as real dependencies) is to create a custom Docker image. I'm not a huge fan of this approach, unless custom image is what you release for your (on-prem) users.

As proved by the code, starting and initialising Elasticsearch with data from an automated test is doable.
Compared to other dependencies available as [TC modules](https://testcontainers.com/modules/), ES is more challenging to init with data. The reason is lack of "init directory in Docker container" (e.g. `/docker-entrypoint-initdb.d` for MySQL), but I understand why it's not there and I'm not sure it should be added, because reasons.
Also, at least from Java stack PoV, there's nothing to "keep your stuff up to date and migrate if necessary", like Flyway or Liquibase. Again, such a technique might not fit well, mentally or technically.
The thing is: people think using terms and concepts they already know. For exactly this reason I used in the demo both Elasticsearch and MySQL, to show that (apart from initialising snapshot repo), all the steps in automated tests can be done 1:1, comparing to the technologies people already know.

Perhaps we could add some code, which could help people with initialising and (re)storing data, e.g. in Testcontainer's ES module, or as an external library using it. IMHO working code is far better than any documentation, because the code you run, whereas based on the docs you have to write the code and then run it ;-)

Should anyone be interested in the whole talk, sadly the recording from JCON hasn't been published yet. The best thing I can offer right now (explaining concepts just with MySQL) is from [Devoxx BE](https://youtu.be/bali16KUnOI?t=5202).

Last but not least, I'll vote against any "maintained mock [library]". The only mock-thingy worth considering (which is stack-agnostic) is WireMock, and perhaps this could be IMHO the only thing (semi-)supported or recommended.

From my experience mocks are only useful in fast-detection-network tests, just to check if things are willing to work with each other.
To check if they work _for real_, one needs _real_ dependency, mimicing production, SSL issues, migration hassle, and so on. After first real production disaster people realise they were fooling themselves with (incorectly used or maintained) mocks. Because a mock is good only when it reflects the behaviour of the thing it mocks, with all the documented (and undocumented!) changelog adaptations.

Tests are needed for confidence before production, not to show that "look, my tests are running in 5 seconds". (And there are also ways of making integration tests run fast, going from minutes to seconds, demonstrated in the mentioned talk.)

Having all that written (sorry! :-)), I'm happy to help with anything that could led Elasticsearch users towards better and easier tests.

**Comment by lustickd** [comment]
For anyone else coming here from the [Unit Tests](https://www.elastic.co/guide/en/elasticsearch/reference/8.18/unit-tests.html) page, if you are running Elasticsearch on Github Actions I eventually got this configuration to work:

```yml
    services:
      elasticsearch:
        image: elasticsearch:8.17.5
        ports: ['9200:9200']
        env:
          action.destructive_requires_name: false
          discovery.type: single-node
          xpack.security.enabled: false
        options: >-
          --health-cmd "curl --silent --fail localhost:9200/_cluster/health"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
```

Locally I use my dev elasticsearch setup, make sure to prefix your index names with "test" in your setup so you don't accidentally delete your development indexes.
